File: cpp/src/Matrix.cpp
#include "Matrix.h"
using namespace std;

ZMAT_BEGIN_NAMESPACE

Matrix::Matrix(const Complex &value)
{
    this->_rowNum = 0;
    this->_colNum = 0;
    this->setSize(1, 1);
    this->setData(1, 1, value);
}

Matrix::Matrix(const int rowNum, const int colNum)
{
    this->_rowNum = 0;
    this->_colNum = 0;
    this->setSize(rowNum, colNum);
}

Matrix::~Matrix()
{
}

const int Matrix::rowNum() const
{
    return this->_rowNum;
}

const int Matrix::colNum() const
{
    return this->_colNum;
}

void Matrix::setRowNum(const int rowNum)
{
    this->setSize(rowNum, this->_colNum);
}

void Matrix::setColNum(const int colNum)
{
    this->setSize(this->_rowNum, colNum);
}

void Matrix::setSize(const int rowNum, const int colNum)
{
    vector< vector<Complex> > data;
    for (int i = 0; i < rowNum; ++i)
    {
        data.push_back(vector<Complex>());
        for (int j = 0; j < colNum; ++j)
        {
            if (i < this->_rowNum && j < this->_colNum)
            {
                data[i].push_back(this->_data[i][j]);
            }
            else
            {
                data[i].push_back(this->_zero);
            }
        }
    }
    this->_data = data;
    this->_rowNum = rowNum;
    this->_colNum = colNum;
}

const Matrix Matrix::row(const int r) const
{
    Matrix R(1, this->_colNum);
    for (int i = 1; i <= this->_colNum; ++i)
    {
        R(1, i) = this->data(r, i);
    }
    return R;
}

const Matrix Matrix::col(const int c) const
{
    Matrix C(this->_rowNum, 1);
    for (int i = 1; i <= this->_rowNum; ++i)
    {
        C(i, 1) = this->data(i, c);
    }
    return C;
}

Complex& Matrix::operator ()(const int r, const int c)
{
    if (r >= 1 && r <= this->_rowNum)
    {
        if (c >= 1 && c <= this->_colNum)
        {
            return this->_data[r - 1][c - 1];
        }
    }
    return this->_zero;
}

const Complex Matrix::data(const int r, const int c) const
{
    if (r >= 1 && r <= this->_rowNum)
    {
        if (c >= 1 && c <= this->_colNum)
        {
            return this->_data[r - 1][c - 1];
        }
    }
    return this->_zero;
}

void Matrix::setData(const int r, const int c, const Complex &value)
{
    if (r >= 1 && r <= this->_rowNum)
    {
        if (c >= 1 && c <= this->_colNum)
        {
            this->_data[r - 1][c - 1] = value;
        }
    }
}

ZMAT_END_NAMESPACE

File: js/src/matrix.js
/**
 * The matrix.
 * The indices start from 1 in the matrix.
 * @author ZHG <CyberZHG@gmail.com>
 */
var Matrix = {
    /**
     * Create a new matrix object filled with complex 0.
     * @param {int} rowNum The number of rows of the matrix.
     * @param {int} colNum The number of columns of the matrix.
     * @return {matrix} The matrix object.
     */
    New: function(rowNum, colNum) {
        rowNum = arguments[0] ? arguments[0] : 1;
        colNum = arguments[1] ? arguments[1] : 1;
        var matrix = {
            rowNum: rowNum,
            colNum: colNum,
            data: new Array(rowNum * colNum),
            /**
             * Calculate the position in the data array with the given row and column.
             * @param {int} row The index of the row.
             * @param {int} col The index of the column.
             * @return {int} The position in the data array.
             */
            pos: function(row, col) {
                return (row - 1) * this.colNum + col - 1;
            },
            /**
             * Get data from matrix.
             * @param {int} row The index of the row.
             * @param {int} col The index of the column.
             * @return {complex} The value of the matrix, returns complex 0 if any index is illegal.
             */
            get: function(row, col) {
                if (row >= 1 && row <= this.rowNum) {
                    if (col >= 1 && col <= this.colNum) {
                        return this.data[this.pos(row, col)];
                    }
                }
                return Complex.New();
            },
            /**
             * Set data to matrix.
             * @param {int} row The index of the row.
             * @param {int} col The index of the column.
             */
            set: function(row, col, value) {
                if (row >= 1 && row <= this.rowNum) {
                    if (col >= 1 && col <= this.colNum) {
                        if (typeof(value) == "number") {
                            value = Complex.New(value);
                        }
                        this.data[this.pos(row, col)] = value;
                    }
                }
            },
            /**
             * Add with another matrix.
             * @param {matrix} Another matrix.
             * @return {matrix} The added result.
             */
            add: function(mat) {
                var res = Matrix.New(this.rowNum, this.colNum);
                for (var i = 1; i <= this.rowNum; ++i) {
                    for (var j = 1; j <= this.colNum; ++j) {
                        res.set(i, j, this.get(i, j).add(mat.get(i, j)));
                    }
                }
                return res;
            },
            /**
             * Subtract to another matrix.
             * @param {matrix} Another matrix.
             * @return {matrix} The subtracted result.
             */
            sub: function(mat) {
                var res = Matrix.New(this.rowNum, this.colNum);
                for (var i = 1; i <= this.rowNum; ++i) {
                    for (var j = 1; j <= this.colNum; ++j) {
                        res.set(i, j, this.get(i, j).sub(mat.get(i, j)));
                    }
                }
                return res;
            },
            /**
             * Multiply with another matrix.
             * @param {matrix} Another matrix.
             * @return {matrix} The multiplied result.
             */
            mul: function(mat) {
                var res = Matrix.New(this.rowNum, mat.colNum);
                for (var i = 1; i <= this.rowNum; ++i) {
                    for (var j = 1; j <= mat.colNum; ++j) {
                        for (var k = 1; k <= this.colNum; ++k) {
                            res.set(i, j, res.get(i, j).add(this.get(i, k).mul(mat.get(k, j))));
                        }
                    }
                }
                return res;
            },
            /**
             * Whether the matrix equals to another.
             * @param {matrix} mat Another matrix.
             * @return {boolean} Returns true if equals, otherwise false.
             */
            equal: function(mat) {
                for (var i = 1; i <= this.rowNum; ++i) {
                    for (var j = 1; j <= this.colNum; ++j) {
                        if (!this.get(i, j).equal(mat.get(i, j))) {
                            return false;
                        }
                    }
                }
                return true;
            },
            /**
             * Returns a string represents the complex.
             * @return {string} The string
             */
            str: function() {
                var res = "[";
                for (var i = 1; i <= this.rowNum; ++i) {
                    for (var j = 1; j <= this.colNum; ++j) {
                        res += " " + this.get(i, j).str();
                    }
                    if (i != this.rowNum) {
                        res += ";";
                    }
                }
                res += "]";
                return res;
            }
        };
        for (var i = 1; i <= rowNum; ++i) {
            for (var j = 1; j <= colNum; ++j) {
                matrix.set(i, j, Complex.New());
            }
        }
        return matrix;
    }
};